lib/customisation.nix

```diff
@@ -306,4 +310,72 @@ rec {
       };
     in self;
 
+  /* Define a `mkDerivation`-like function based on another `mkDerivation`-like function.
+
+  `mkDerivation` gives access to its final set of derivation attributes when it is passed a function,
+  or when it is passed an overlay-style function in `overrideAttrs`.
+  Instead of composing new `mkDerivation`-like builders using normal function composition,
+  `extendMkDerivation` makes sure that the returned builder supports such
+  first class recursion like `mkDerivation` does.
+
+     Type:
+       extendMkDerivation ::
+         ((AttrSet -> AttrSet) -> Derivation)
+         -> (AttrSet -> AttrSet -> AttrSet)
+         -> (((AttrSet -> AttrSet) | AttrSet) -> Derivation)
+
+     Example:
+       mkLocalDerivation = lib.extendMkDerivation pkgs.stdenv.mkDerivation (finalAttrs:
+         args@{ preferLocalBuild ? true, allowSubstitute ? false, ... }:
+         args // { inherit preferLocalBuild allowSubstitute; })
+
+       mkLocalDerivation.__functionArgs
+       => { allowSubstitute = true; preferLocalBuild = true; }
+
+       mkLocalDerivation { inherit (pkgs.hello) pname version src; }
+       => «derivation /nix/store/fh9jww7kv5v814lasn78ydck7fb6hsjr-hello-2.12.1.drv»
+
+       mkLocalDerivation (finalAttrs: { inherit (pkgs.hello) pname version src; })
+       => «derivation /nix/store/fh9jww7kv5v814lasn78ydck7fb6hsjr-hello-2.12.1.drv»
+  **/
+  extendMkDerivation =
+    # mkDerivation-like builder to extend
+    mkDerivationBase:
+    # Overlay implementation of the new builder
+    impl:
+    lib.copyFunctionArgs
+      # Make the __functionArgs looks like one from a builder accepting plain attribute set.
+      (impl { })
+      # Adds the fixed-point style support
+      (rattrs: (mkDerivationBase (finalAttrs:
+        impl finalAttrs (if lib.isFunction rattrs then rattrs finalAttrs else rattrs)
+      )));
```

Comment:
This needs tests.

 - Overriding up to `mkDerivation` is tested in `pkgs/test/overriding.nix`. This function is not language specific, so it could be tested there too.

 - Other than that, a couple of languages have their own files in that directory. You can't let Python be beat by PHP, no? ;) I'm sure you can think of some examples that test how the python logic works, e.g. invocations of `extendMkDerivation` or `extendMkDerivationModified.

Reply:
I'll try to add some.

Reply:
Now `.overridePythonAttrs`, `.override` and `.override .overridePythonAttrs` are all covered by unit tests.